File: Clients/JavaQGl/Q/src/com/qframework/core/GLModel.java
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

package com.qframework.core;


import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.ShortBuffer;
import java.util.Iterator;
import java.util.ArrayList;
import javax.media.opengl.GL;
import javax.media.opengl.GL2;


public class GLModel {

	protected boolean mEnabled = false;
	protected boolean mForceHalfTexturing = false;
	protected int     mForcedOwner = 0;
	private ArrayList<GLShape>	mShapeList = new ArrayList<GLShape>();	
	private ArrayList<GLVertex>	mVertexList = new ArrayList<GLVertex>();
	private int mIndexCount = 0;
    private FloatBuffer   mVertexBuffer;
    private ByteBuffer   mColorBuffer;
    private ShortBuffer mIndexBuffer;
    private FloatBuffer   mTextureBuffer;
    private int			mTextureOffset = 0;
    private int			mTextureW;
    private int			mTextureH;

    protected int mTextureID = 1;
    
    protected boolean mTransform = false;
    
    protected float mBBoxMin[];
    protected float mBBoxMax[];
    protected int mVertexOffset = 0;
	private boolean mGenerated = false;
	protected String mName;
    protected GameonApp mApp;
    
	public GLModel(String name , GameonApp app) {
		mApp  = app;
		mName = name;
		mBBoxMin = new float[3];
	    mBBoxMax = new float[3];
	    mBBoxMin[0] = 1E10f;
	    mBBoxMin[1] = 1E10f;
	    mBBoxMin[2] = 1E10f;
	    mBBoxMax[0] = -1E10f;
	    mBBoxMax[1] = -1E10f;
	    mBBoxMax[2] = -1E10f;	
	    
    	mTextureOffset = 0;
        mTextureW = 1;
        mTextureH = 1;
    	
	}
	public boolean isValid()
	{
		return mVertexList.size() > 0;
	}
	public void addShape(GLShape shape) {
		mShapeList.add(shape);
		mIndexCount += shape.getIndexCount();
	}
	
	public void generate() {
        mGenerated = true;
        ByteBuffer bb = ByteBuffer.allocateDirect(mVertexList.size()*4*4);
        bb.order(ByteOrder.nativeOrder());
        mColorBuffer = bb;

        bb = ByteBuffer.allocateDirect(mVertexList.size()*3*Float.SIZE /8);
        bb.order(ByteOrder.nativeOrder());
        mVertexBuffer = bb.asFloatBuffer();

        bb = ByteBuffer.allocateDirect(mIndexCount* Short.SIZE/8);
        bb.order(ByteOrder.nativeOrder());
        mIndexBuffer = bb.asShortBuffer();

        bb = ByteBuffer.allocateDirect(mVertexList.size()*Float.SIZE/8 * 2 * (mTextureW * mTextureH));
        bb.order(ByteOrder.nativeOrder());
        mTextureBuffer = bb.asFloatBuffer();

		if (mTextureW == 1 && mTextureH == 1) {
			Iterator<GLVertex> iter2 = mVertexList.iterator();
			while (iter2.hasNext()) {
				GLVertex vertex = iter2.next();
				vertex.put(mVertexBuffer, mColorBuffer, mTextureBuffer);
			}
		} else {
			Iterator<GLVertex> iter2 = mVertexList.iterator();
			while (iter2.hasNext()) {
				GLVertex vertex = iter2.next();
				vertex.put(mVertexBuffer, mColorBuffer);
			}
			
			int cnt = 0;
			int cnttrig = mVertexList.size() /2;
			int fx = mForcedOwner % mTextureW;
			int fy = mForcedOwner / mTextureW;
			for (int b=0; b< mTextureH; b++) {
				for (int a=0; a< mTextureW; a++) {
					cnt = 0;
					Iterator<GLVertex> iter3 = mVertexList.iterator();
					while (iter3.hasNext()) {
						GLVertex vertex = iter3.next();
						if (mForceHalfTexturing && cnt >= cnttrig) 
						{
							vertex.putText( mTextureBuffer , fx, fy,  mTextureW, mTextureH);
						}else
						{
							vertex.putText( mTextureBuffer , a, b,  mTextureW, mTextureH);
						}
						cnt++;
					}					
					
				}
			}
		}

		Iterator<GLShape> iter3 = mShapeList.iterator();
		while (iter3.hasNext()) {
			GLShape shape = iter3.next();
			shape.putIndices(mIndexBuffer);
		}
		
	}
	
	
	public GLVertex addVertex(float x, float y, float z, float tu, float tv) {
		GLVertex vertex = new GLVertex(x, y, z, tu , tv , mVertexList.size());
		mVertexList.add(vertex);
		if (x < mBBoxMin[0]) mBBoxMin[0] = x;
		if (y < mBBoxMin[1]) mBBoxMin[1] = y;
		if (z < mBBoxMin[2]) mBBoxMin[2] = z;		

		if (x > mBBoxMax[0]) mBBoxMax[0] = x;
		if (y > mBBoxMax[1]) mBBoxMax[1] = y;
		if (z > mBBoxMax[2]) mBBoxMax[2] = z;

		return vertex;
	}	

	int count = 0;
	static int mLastText = -1;
	public void setupRef(GL2 gl)
	{
		if (mTextureID != mLastText || mApp.textures().isUpdated())
		{
			if (gl.glIsTexture(mTextureID))
			{
				gl.glBindTexture(GL2.GL_TEXTURE_2D, mTextureID);
				mLastText = mTextureID;
				
			}else
			{
				gl.glBindTexture(GL2.GL_TEXTURE_2D, 1);
				mLastText = 1;
				
			}
			
			mApp.textures().resetUpdated();
		}
	}
	public boolean drawRef(GL2 gl, GameonModelRef ref , boolean initRef)
	{
    	if (!mEnabled || mIndexCount == 0) {
    		return true;
    	}

    	
    	if (ref.animating())
    	{
    		ref.animate(mApp.frameDelta());
    	}
    	if (!ref.getVisible() || !ref.mEnabled){
    		return true;
    	}
    	if (initRef)
    	{
    		setupRef(gl);
    	}
    	

    	
    	if (ref.mTransformOwner) {
    		if (ref.mOwner >=0 && ref.mOwner < ref.mOwnerMax)
    			mTextureOffset = mTextureBuffer.capacity() * ref.mOwner / ref.mOwnerMax;
    		else
    			System.out.println( "owner error " + ref.mOwner);
    	}


		mColorBuffer.position(0);
		mVertexBuffer.position(mVertexOffset);
        mTextureBuffer.position(mTextureOffset);
        mIndexBuffer.position(0);

        if (mColorBuffer.remaining() == 0)
        {
        	System.out.println("canceling " + mName);
        	return false;
        }
        gl.glVertexPointer(3, GL.GL_FLOAT, 0, mVertexBuffer);
        gl.glColorPointer(4, GL.GL_UNSIGNED_BYTE, 0, mColorBuffer);
        gl.glTexCoordPointer(2, GL.GL_FLOAT, 0, mTextureBuffer);
	
    	gl.glPushMatrix();
    	gl.glMultMatrixf( ref.matrix() , 0);
        if (mIndexCount > 0)
        {
        	gl.glDrawElements(GL.GL_TRIANGLES, mIndexCount, GL.GL_UNSIGNED_SHORT, mIndexBuffer);
        }

    	
   		gl.glPopMatrix();

   		return false;
   	
	}
   

    public void setTextureOffset(int w, int h) {
    	mTextureW = w;
    	mTextureH = h;
    	mTextureOffset = 0;
    }
    
	public void reset() {
		mColorBuffer.clear();
	    mVertexBuffer.clear();
	    mIndexBuffer.clear();
		mTextureBuffer.clear();
		generate();

	}

	public void setupOffset(int value)
	{
	    mVertexOffset = value * 12;
	}

	public void forceIndexCount( int count)
	{
		mIndexCount = count;
	}
}

File: Clients/iPhoneQ/Q/QFramework/GLModel.m
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

#import "GLModel.h"
#import "ByteBuffer.h"
#import "FloatBuffer.h"
#import "ShortBuffer.h"
#import "GLShape.h"
#import "GLVertex.h"
#import "GameonModelRef.h"
#import "TextureFactory.h"
#import "GameonApp.h"

#import <OpenGLES/ES1/gl.h>
#import <OpenGLES/ES1/glext.h>

@implementation GLModel

@synthesize mEnabled;
@synthesize mTextureID;
@synthesize mForceHalfTexturing;
@synthesize mForcedOwner;

- (id) initWithApp:(GameonApp*)app
{
	if (self = [super init])
	{
        mVertexList = [[NSMutableArray alloc] init] ;
        mShapeList = [[NSMutableArray alloc] init] ;
        mApp = app;
        mEnabled = false;
        mTextureOffset = 0;
        mTextureW = 1;
        mTextureH = 1;        
        
        mBBoxMin[0] = 1E10f;
        mBBoxMin[1] = 1E10f;
        mBBoxMin[2] = 1E10f;
        mBBoxMax[0] = -1E10f;
        mBBoxMax[1] = -1E10f;
        mBBoxMax[2] = -1E10f;	
        
        
        mIndexCount = 0;
        
        mVertexBuffer = [[FloatBuffer alloc] init] ;
        mColorBuffer = [[ByteBuffer alloc] init] ;
        mIndexBuffer = [[ShortBuffer alloc] init] ;
        mTextureBuffer = [[FloatBuffer alloc] init] ;

        mTextureOffset = 0;
        mTextureW = 1;
        mTextureH = 1;
        mTextureID = 1;
        mTransform = false;
        mForceHalfTexturing = false;
        mForcedOwner = 0;
        mVertexOffset = 0;
        
    }
    return self;
}

- (void) dealloc 
{
    [mVertexList release];
    [mShapeList release] ;
    [mVertexBuffer release];
    [mColorBuffer release];
    [mIndexBuffer release];
    [mTextureBuffer release];
    
    [super dealloc];  
}

- (void) addShape:(GLShape*) shape 
{
    [mShapeList addObject:shape];
    mIndexCount += [shape getIndexCount];
}




- (void) generate 
{	

    if (mTextureW == 1 && mTextureH == 1) {
        for (int a=0; a< [mVertexList count]; a++)
        {
            GLVertex* vertex = [mVertexList objectAtIndex:a];
            [vertex put:mVertexBuffer cb:mColorBuffer tb:mTextureBuffer];
        }
        
    } else {
        for (int a=0; a< [mVertexList count]; a++)
        {
            GLVertex* vertex = [mVertexList objectAtIndex:a];
            [vertex put:mVertexBuffer cb:mColorBuffer];
        }
        
        int cnt = 0;
        int cnttrig = [mVertexList count] /2;
        int fx = mForcedOwner % mTextureW;
        int fy = mForcedOwner / mTextureW;
        
        for (int b=0; b< mTextureH; b++) {
            for (int a=0; a< mTextureW; a++) {
                cnt = 0;

                for (int c = 0; c< [mVertexList count]; c++)
                {
                    GLVertex* vertex = [mVertexList objectAtIndex:c];
                    if (mForceHalfTexturing && cnt >= cnttrig) 
                    {
                        [vertex putText:mTextureBuffer x:fx y:fy w:mTextureW h:mTextureH];                        
                    }else
                    {
                        [vertex putText:mTextureBuffer x:a y:b w:mTextureW h:mTextureH];
                    }
                    cnt++;
                    

                }
            }
        }
    }
    for (int a=0; a< [mShapeList count]; a++)
    {
        GLShape* shape = [mShapeList objectAtIndex:a];
        [shape  putIndices:mIndexBuffer];
        
    }
    
    
}

-(GLVertex*) addVertex:(float)ax y:(float)ay z:(float)az tu:(float)atu tv:(float)atv 
{
    GLVertex* vertex = [[[GLVertex alloc] initWithCords:ax y:ay z:az tu:atu tv:atv i:[mVertexList count]] autorelease];

    [mVertexList addObject:vertex];
    if (ax < mBBoxMin[0]) mBBoxMin[0] = ax;
    if (ay < mBBoxMin[1]) mBBoxMin[1] = ay;
    if (az < mBBoxMin[2]) mBBoxMin[2] = az;		
    
    if (ax > mBBoxMax[0]) mBBoxMax[0] = ax;
    if (ay > mBBoxMax[1]) mBBoxMax[1] = ay;
    if (az > mBBoxMax[2]) mBBoxMax[2] = az;
    
    return vertex;
}	

- (void) setupRef
{
    
    static int lasttextId = -1;
//    glVertexPointer(3, GL_FLOAT, 0, [mVertexBuffer get]);
//    glColorPointer(4, GL_UNSIGNED_BYTE, 0, [mColorBuffer get]);
    
    if (mTextureID != lasttextId || [mApp.textures isUpdated])
    {
		if (glIsTexture(mTextureID) == GL_TRUE)
		{
			glBindTexture(GL_TEXTURE_2D, mTextureID);
			lasttextId = mTextureID;
		}else
		{
            NSLog(@" bad texture %d "  , mTextureID); 
			glBindTexture(GL_TEXTURE_2D, 1);
			lasttextId = 1;
		}
		[mApp.textures resetUpdate];
    }
    //gl.glTexCoordPointer(2, GL10.GL_FIXED, 0, mTextureBuffer);		
}

-(void) drawRef:(GameonModelRef*)ref init:(bool)initRef
{
    if (!mEnabled || mIndexCount ==0 ) {
        return;
    }
    
	if (ref.animating)
	{
		[ref animate:mApp.frameDelta];
	}
	if (![ref getVisible] || !ref.mEnabled)
	{
		return;
	}
	if (initRef)
	{
		[self setupRef];
	}

		
    if (ref.mTransformOwner) {
        if (ref.mOwner >=0 && ref.mOwner < ref.mOwnerMax)
            mTextureOffset = [mTextureBuffer capacity] * ref.mOwner / ref.mOwnerMax;	
    }else {
        mTextureOffset = 0;
    }
    
    glPushMatrix();

    /*
    NSLog(@" %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f  %f ",
          ref.mMatrix[0],ref.mMatrix[1],ref.mMatrix[2],ref.mMatrix[3],
          ref.mMatrix[4],ref.mMatrix[5],ref.mMatrix[6],ref.mMatrix[7],
          ref.mMatrix[8],ref.mMatrix[9],ref.mMatrix[10],ref.mMatrix[11],
          ref.mMatrix[12],ref.mMatrix[13],ref.mMatrix[14],ref.mMatrix[15]);          
    */
    glMultMatrixf( ref.mMatrix);
    float* toffset = [mTextureBuffer get] + mTextureOffset;
    glTexCoordPointer(2, GL_FLOAT, 0, toffset);
    
    glVertexPointer(3, GL_FLOAT, 0, [mVertexBuffer get]);
    glColorPointer(4, GL_UNSIGNED_BYTE, 0, [mColorBuffer get]);
    
    //glBindTexture(GL_TEXTURE_2D, mTextureID);
    glDrawElements(GL_TRIANGLES, mIndexCount, GL_UNSIGNED_SHORT, [mIndexBuffer get]);    
    
    glPopMatrix();
    
    
}

-(void) setTextureOffset:(int)aw h:(int)ah 
{
    mTextureW = aw;
    mTextureH = ah;
    mTextureOffset = 0;
}
- (void) reset {
    //generate;
    
}

- (void) setupOffset:(int) value
{
    mVertexOffset = value * 12;
    glVertexPointer(3, GL_FLOAT, 0, [mVertexBuffer get:mVertexOffset]);    
    glColorPointer(4, GL_UNSIGNED_BYTE, 0, [mColorBuffer get:value * 16]);
}

-(void) forceIndexCount:(int)count
{
    mIndexCount = count;
}

@end
